### ISSUE TITLE
Provide assistance for writing an impl block

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust-analyzer#17590]
When I am at a type, I can select a "generate impl" assist and it will generate a block right below the type. That's useful, but often I am somewhere else in the project (might not even be the same file), and I want to generate an impl block here. Would be great if there was some way to get the assist there as well. :)

What I usually try is to write `impl TypeName`, and then my hope would be that (a) I get auto-complete (including auto-import) for the type name, and (b) I can then select an assist to generate an `impl` block, and it will turn this into
```rust
impl<'a, T: Bound> TypeName<'a, T> {
  /* cursor here */
}
```

**Comment by lnicola** [comment]
A better UX might be to write `impl TypeName` and get a completion for the whole thing :slightly_smiling_face:.

**Comment by ShoyuVanilla** [comment]
I'll give this a try
@rustbot claim

**Comment by ChayimFriedman2** [comment]
@ShoyuVanilla Are you still working on this?

**Comment by ShoyuVanilla** [comment]
No, I was just waiting for #18032 to effectively work on assist things. You could work on this if you like 😄

**Comment by ChayimFriedman2** [comment]
@rustbot release-assignment

I still intend to work on this, but as a diagnostic with quick fix, and this diagnostic requires https://github.com/rust-lang/rust-analyzer/pull/18074 and follow-ups, which is quite big work. So in order to not create false hopes and not block anyone who would like to want on this, I'm unassigning for now.

**Comment by ChayimFriedman2** [comment]
Now after we have diagnostics for type, I can come back to work on this.

@rustbot claim
